Embed block: use preview html for flickr embed previews in editor

mashikag opened the pull request:
## What?
<!-- In a few words, what is the PR actually doing? -->
This commit is fixing an issue where flickr embeds are displayed with low quality images in the block editor.

The issue was reported [here](https://github.com/WordPress/gutenberg/issues/11962).

<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
The issue was that thumbnail image (low resolution image) was used for flickr's embed previews. On top of that the preview was stretched to 100% width of the parent element, which most of the time meant more than width of the preview's thumbnail image. This resulted in flickr embed previews appear "pixelated" in the editor. 
Flickr oEmbed previews returned by their oEmbed API are marked as 'photo' types. According to our source code we prefer to use the thumbnail url, over preview's html when embed's type is 'photo'.
![Screenshot 2022-04-08 at 14 25 48](https://user-images.githubusercontent.com/2019970/162435304-bb78b88e-8060-4c4b-a1f5-5eb6d3e479a5.png)

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
I fixed this by introducing new block attribute of 'preferHtml' which, when set to `true`, forces the use of preview's html even when an embed is of 'photo' type.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
1. Open Post or Page editor.
2. Copy any flickr share url of your liking, e.g. `https://flic.kr/p/oj7o7q`
3. Paste it into your block editor.
4. After the fix the preview should not appear to be a "pixelated" image, it should look the same way as when you preview the page with the embed in it.

## Screenshots or screencast <!-- if applicable -->

#### Before fix
![flickr-before-fix](https://user-images.githubusercontent.com/2019970/162438298-7d29dc2d-7a53-4068-aa7e-917140491126.gif)

#### After fix
![flickr-after-fix](https://user-images.githubusercontent.com/2019970/162438330-a411acc1-976d-4a06-bbf9-57fe820dd3ed.gif)

mashikag commented:
Despite the above initial fix implementation, I wanted to ask people reviewing this whether they can justify the need for `getPhotoHtml()` function (see [code](https://github.com/mashikag/gutenberg/blob/89109821319e4e672be0ab181b045febb86cbe15/packages/block-library/src/embed/util.js#L68)).

![Screenshot 2022-04-08 at 14 57 51](https://user-images.githubusercontent.com/2019970/162441157-9519fb9e-4b4d-4e1d-9cb5-9867845642cf.png)
 
I tried to look for other examples than `flickr` which would return embeds with `photo` type. I could not find any. 

1. My worry is that the thumbnail preview is broken for all `photo` type embed previews
2. and I just cannot justify to myself ever using the thumbnail url, over preview's html - given both of them are available.

ntsekouras reviewed (commented):
Thanks for the PR @mashikag!

I took a look here and I was wondering if it would be simpler to just change [this line of code](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/embed/util.js#L71) and give priority to the full image. Would that work?

mashikag commented:
> I took a look here and I was wondering if it would be simpler to just change [this line of code](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/embed/util.js#L71) and give priority to the full image. Would that work?

Yes, it does work. However would that not break any other `photo` type embeds, that are not flickr? I do not know myself if any other supported embeds provide `photo` type previews? 

It would make even more sense to me to always use `html` provided with preview, even if the preview's `type` is `photo`. This could be done by rewriting [this line](https://github.com/WordPress/gutenberg/blob/bd2497e2899f40ff4044c59640f6a980663090a0/packages/block-library/src/embed/embed-preview.js#L71) as:
```
const html =  preview.html || (type === 'photo' && getPhotoHtml( preview )) ;
```
Or even rename the `getPhotoHtml` to `getHtml` or `createHtml`, and let it handle a case where `html` is not provided for any embed type, by using its current internal logic, like so:
```
const html =  preview.html || createHtml( preview );
```
 But it just makes me question what was the purpose for this function when it was written, which embedded service (live example) required this handling? Would you have any insight to this @ntsekouras ?

ntsekouras commented:
>Yes, it does work. However would that not break any other photo type embeds, that are not flickr? I do not know myself if any other supported embeds provide photo type previews?

I'm not sure how this could break things. It was actually used before in code and if the `photo.url` doesn't exist, it will fallback to the thumbnail url.

>It would make even more sense to me to always use html provided with preview, even if the preview's type is photo. 

>But it just makes me question what was the purpose for this function when it was written, which embedded service (live example) required this handling? Would you have any insight to this @ntsekouras ?

I'm not really sure about the reasoning of the original coding of this and I'm not sure if your suggested change will affect other embedded previews. That's why I suggested the above change of priority - it was used before and gives us the wanted result.

mashikag commented:
> I'm not sure how this could break things. It was actually used before in code and if the photo.url doesn't exist, it will fallback to the thumbnail url.

Yes, it was used but only if the `photo.thumbnail_url` did not exist. 😄  I am just worried that for some `photo` type embed scenario, for which the function was originally written, using `photo.url` over `photo.thumbnail_url` will result in an unexpected user experience, just like using the `photo. thumbnail_url` does for flickr embed.

On the other hand, I could just leave a note beside the change, linking to the issue, in case it will break something as per above.

mashikag committed:
Embed block: use embed url over thumbnail_url for html of photo type embeds

This commit is fixing an issue where flickr embeds display pixelated preview images in the block editor.

The issue was that flickr embed (variation of embed block) preview used thumbnail image, over better quality preview options, such as full image url. The thumbnail image is low resolution and 1:1 aspect ratio (often cropped out) version of the original image. Then the preview is stretched to 100% width of the parent element. This resulted in low resolution image being stretched out beyond its width and height. Hence it appeared pixelated.

The issue is fixed by using full image `url`, over `thumbnail_url`, when generating html for `photo` type embed previews.

mashikag commented:
Hey @ntsekouras ! 👋   Could you have a look at the PR again, please? 🙏

ntsekouras commented on packages/block-library/src/embed/util.js:
```diff
@@ -66,9 +66,11 @@ export const isFromWordPress = ( html ) =>
 	html && html.includes( 'class="wp-embedded-content"' );
 
 export const getPhotoHtml = ( photo ) => {
+	// If full image url not found use thumbnail_url: https://github.com/WordPress/gutenberg/pull/40187
```
```suggestion
	// If full image url not found use thumbnail.
```

No need to link the PR in comments. It can be found through git history.

ntsekouras committed:
Update packages/block-library/src/embed/util.js

ntsekouras reviewed (approved):
Thanks for your work @mashikag!  Let's try this.

mashikag commented:
Thanks @ntsekouras ! 🙇 

BTW just letting you know because I am not sure if it is obvious or not, but I do not have merge rights here.

ntsekouras commented:
>BTW just letting you know because I am not sure if it is obvious or not, but I do not have merge rights here.

I knew that, yes 😄 . I was waiting for the CI to go 🟢 . Thanks again!
